ztools/ztls/handshake_client.go

```diff
@@ -32,19 +32,246 @@ type clientHandshakeState struct {
 	session         *ClientSessionState
 }
 
+type CacheKeyGenerator interface {
+	Key(net.Addr) string
+}
+
+type ClientHelloConfiguration struct {
+	//Version in the handshake header
+	HandshakeVersion uint16
+
+	//if len == 32, it will specify the client random
+	//Otherwise, the field will be random
+	ClientRandom []byte
+
+	//if RandomSessionID > 0, will overwrite SessionID w/ that many
+	//random bytes when a session resumption occurs
+	RandomSessionID int
+	SessionID       []byte
+
+	//These fields will appear exactly in order in the ClientHello
+	CipherSuites       []uint16
+	CompressionMethods []uint8
+	Extensions         []ClientExtension
+
+	//Optional, both must be non-nil, or neither.
+	//Custom Session cache implementations allowed
+	SessionCache ClientSessionCache
+	CacheKey     CacheKeyGenerator
+}
+
+func (c *ClientHelloConfiguration) ValidateExtensions() error {
+	for _, ext := range c.Extensions {
+		switch ext.(type) {
+		case PointFormatExtension:
```

Comment:
Most extensions don't need to make sure their details are implemented, so the default is no action. Validate may not be the best wording. Also, maybe this is best rolled into another function in the Extension interface, and calling it on all extensions here?